File: plugin/preset/elements/span.lua
local Span = {}

function Span:new(content)
	local o = {
		content = content,
		foreground = "",
		background = "",
		modifier = 0,
	}
	setmetatable(o, self)
	self.__index = self
	return o
end

function Span:fg(color)
	self.foreground = color
	return self
end

function Span:bg(color)
	self.background = color
	return self
end

function Span:bold()
	self.modifier = self.modifier | 1
	return self
end

function Span:dim()
	self.modifier = self.modifier | 2
	return self
end

function Span:italic()
	self.modifier = self.modifier | 4
	return self
end

function Span:underline()
	self.modifier = self.modifier | 8
	return self
end

function Span:blink()
	self.modifier = self.modifier | 16
	return self
end

function Span:blink_rapid()
	self.modifier = self.modifier | 32
	return self
end

function Span:reverse()
	self.modifier = self.modifier | 64
	return self
end

function Span:hidden()
	self.modifier = self.modifier | 128
	return self
end

function Span:crossed()
	self.modifier = self.modifier | 256
	return self
end

function Span:reset()
	self.foreground = ""
	self.background = ""
	self.modifier = 0
	return self
end

function Span:style(style)
	if style.fg then
		self:fg(style.fg)
	end
	if style.bg then
		self:bg(style.bg)
	end
	if style.bold then
		self:bold()
	end
	if style.dim then
		self:dim()
	end
	if style.italic then
		self:italic()
	end
	if style.underline then
		self:underline()
	end
	if style.blink then
		self:blink()
	end
	if style.blink_rapid then
		self:blink_rapid()
	end
	if style.reverse then
		self:reverse()
	end
	if style.hidden then
		self:hidden()
	end
	if style.crossed then
		self:crossed()
	end
	return self
end

function Span:to_string()
	return string.format("%s,%s,,%s;%s", self.foreground, self.background, self.modifier, self.content)
end

setmetatable(Span, {
	__call = function(self, content) return self:new(content) end,
	__tostring = function(self) return self:to_string() end,
})

yazi = yazi or {}
yazi.Span = Span

File: yazi-plugin/src/layout/span.rs
use yazi_config::theme::Color;
use mlua::{AnyUserData, FromLua, Lua, Table, UserData, UserDataMethods, Value};

use super::Style;
use crate::{GLOBALS, LUA};

#[derive(Clone)]
pub(crate) struct Span(pub(super) ratatui::text::Span<'static>);

impl Span {
	pub(crate) fn install() -> mlua::Result<()> {
		let ui: Table = GLOBALS.get("ui")?;
		ui.set(
			"Span",
			LUA.create_function(|_, content: String| Ok(Self(ratatui::text::Span::raw(content))))?,
		)
	}
}

impl<'lua> FromLua<'lua> for Span {
	fn from_lua(value: Value<'lua>, _: &'lua Lua) -> mlua::Result<Self> {
		match value {
			Value::UserData(ud) => Ok(ud.borrow::<Self>()?.clone()),
			_ => Err(mlua::Error::FromLuaConversionError {
				from:    value.type_name(),
				to:      "Span",
				message: Some("expected a Span".to_string()),
			}),
		}
	}
}

impl UserData for Span {
	fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
		methods.add_function("fg", |_, (ud, color): (AnyUserData, String)| {
			ud.borrow_mut::<Self>()?.0.style.fg = Color::try_from(color).ok().map(Into::into);
			Ok(ud)
		});
		methods.add_function("bg", |_, (ud, color): (AnyUserData, String)| {
			ud.borrow_mut::<Self>()?.0.style.bg = Color::try_from(color).ok().map(Into::into);
			Ok(ud)
		});
		methods.add_function("bold", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::BOLD;
			Ok(ud)
		});
		methods.add_function("dim", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::DIM;
			Ok(ud)
		});
		methods.add_function("italic", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::ITALIC;
			Ok(ud)
		});
		methods.add_function("underline", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::UNDERLINED;
			Ok(ud)
		});
		methods.add_function("blink", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::SLOW_BLINK;
			Ok(ud)
		});
		methods.add_function("blink_rapid", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::RAPID_BLINK;
			Ok(ud)
		});
		methods.add_function("hidden", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::HIDDEN;
			Ok(ud)
		});
		methods.add_function("crossed", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier |= ratatui::style::Modifier::CROSSED_OUT;
			Ok(ud)
		});
		methods.add_function("reset", |_, ud: AnyUserData| {
			ud.borrow_mut::<Self>()?.0.style.add_modifier = ratatui::style::Modifier::empty();
			Ok(ud)
		});

		methods.add_function("style", |_, (ud, value): (AnyUserData, Value)| {
			ud.borrow_mut::<Self>()?.0.style = match value {
				Value::Nil => ratatui::style::Style::default(),
				Value::Table(tbl) => Style::from(tbl).0,
				Value::UserData(ud) => ud.borrow::<Style>()?.0,
				_ => return Err(mlua::Error::external("expected a Style or Table or nil")),
			};
			Ok(ud)
		});
	}
}
